homeassistant/components/hyperion/light.py

```diff
@@ -78,34 +110,194 @@
+    hass.async_create_task(
+        migrate_yaml_to_config_entry_and_options(host, port, config[CONF_PRIORITY])
+    )
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up a Hyperion platform from config entry."""
+    host = config_entry.data[CONF_HOST]
+    port = config_entry.data[CONF_PORT]
+    token = config_entry.data.get(CONF_TOKEN)
+
+    async def async_instances_to_entities(response: Dict[str, Any]) -> None:
+        if not response or const.KEY_DATA not in response:
+            return
+        await async_instances_to_entities_raw(response[const.KEY_DATA])
+
+    async def async_instances_to_entities_raw(instances: Dict[str, Any]) -> None:
+        registry = await async_get_registry(hass)
+        entities_to_add: List[Hyperion] = []
+        desired_unique_ids: Set[str] = set()
+        current_entities = hass.data[DOMAIN][config_entry.entry_id][LIGHT_DOMAIN][
+            CONF_ENTITIES
+        ]
+        server_id = config_entry.unique_id
+
+        # In practice, an instance can be in 3 states as seen by this function:
+        #
+        #    * Exists, and is running: Add it to hass.
+        #    * Exists, but is not running: Cannot add yet, but should not delete it either.
+        #      It will show up as "unavailable".
+        #    * No longer exists: Delete it from hass.
+
+        # Add instances that are missing.
+        for instance in instances:
+            instance_id = instance.get(const.KEY_INSTANCE)
+            if instance_id is None or not instance.get(const.KEY_RUNNING, False):
+                continue
+            unique_id = get_hyperion_unique_id(server_id, instance_id)
+            desired_unique_ids.add(unique_id)
+            if unique_id in current_entities:
+                continue
+            hyperion_client = await async_create_connect_client(
+                host, port, instance=instance_id, token=token
+            )
+            if not hyperion_client:
+                continue
+            current_entities.add(unique_id)
+            entities_to_add.append(
+                Hyperion(
+                    unique_id,
+                    instance.get(const.KEY_FRIENDLY_NAME, DEFAULT_NAME),
+                    config_entry.options,
+                    hyperion_client,
+                )
+            )
 
-    async_add_entities([Hyperion(name, priority, hyperion_client)])
+        # Delete instances that are no longer present on this server.
+        for unique_id in current_entities - desired_unique_ids:
+            current_entities.remove(unique_id)
+            async_dispatcher_send(hass, SIGNAL_INSTANCE_REMOVED.format(unique_id))
+            entity_id = registry.async_get_entity_id(LIGHT_DOMAIN, DOMAIN, unique_id)
+            if entity_id:
+                registry.async_remove(entity_id)
+
+        async_add_entities(entities_to_add)
+
+    hass.data[DOMAIN][config_entry.entry_id][LIGHT_DOMAIN] = {CONF_ENTITIES: set()}
```

Comment:
I don't think we even need to store the container in hass.data. The variable should be kept around in memory since it's accessed by a function that we register and store.
```suggestion
    current_entities = set()
```